addons/payment/controllers/portal.py

```diff
@@ -159,9 +159,12 @@ def payment_method(self, **kwargs):
         acquirers_sudo = request.env['payment.acquirer'].sudo()._get_compatible_acquirers(
             request.env.company.id, partner.id, force_tokenization=True, is_validation=True
         )
-        tokens = set(partner.payment_token_ids).union(
-            partner.commercial_partner_id.sudo().payment_token_ids
-        )  # Show all partner's tokens, regardless of which acquirer is available
+
+        # Get all partner's tokens for which acquirers are not disabled.
+        tokens = request.env['payment.token'].search(
+            [('partner_id', 'in', [partner.id, partner.commercial_partner_id.id]),
+             ('acquirer_id', 'in', acquirers_sudo.ids)])
```

Comment:
```suggestion
             ('acquirer_id', 'in', request.env['payment.acquirer'].sudo()._search([
                 ('state', 'in', ['enabled', 'test'])
             ]))
         ])
```

After more thoughts on the problem (cf recent debate on the rd-fun-sales-payment channel), we still want to show tokens from acquirers from other companies (not part of `acquirers_sudo`).
We should limit our fix to a verification on the state of the token's acquirer.

Since searching on `('acquirer_id.state', 'in', ['enabled', 'test'])` wouldn't work because of security restrictions, one available solution is the use of a sudoed subquery to ignore the security rules only for that part of the search domain :).

Reply:
Ok, seems logical to me :).